### PR TITLE
feat(socket): Use OptionalValuePath in the tcp socket config

### PR DESCRIPTION
**Pull request by spencergilbert**
Part of #15635

### CODE REVIEW

File: src/sources/socket/tcp.rs [hunk 1]
```diff
@@ -206,28 +209,29 @@ impl TcpSource for RawTcpSource {
...
 
-                let port_key_path = self.config.port_key.as_ref().map_or_else(
-                    || [BorrowedSegment::from("port")],
-                    |key| [BorrowedSegment::from(key)],
-                );
+                let legacy_port_key = self
+                    .config
+                    .port_key
+                    .as_ref()
+                    .map_or_else(|| parse_value_path("path").ok(), |k| k.path.clone());
```
Comment (neuronull) [inline src/sources/socket/tcp.rs:229]: Shouldn't this be

```suggestion
                    .map_or_else(|| parse_value_path("port").ok(), |k| k.path.clone());
```

? :thinking:

Comment (spencergilbert) [reply src/sources/socket/tcp.rs]: I don't even know what to tell you 🤦

File: src/sources/socket/mod.rs [hunk 2]
```diff
@@ -229,23 +229,20 @@ impl SourceConfig for SocketConfig {
...
                     .as_ref()
-                    .and_then(|x| parse_value_path(x).ok())
                     .map_or_else(
-                        || LegacyKey::InsertIfEmpty(owned_value_path!("host")),
-                        LegacyKey::InsertIfEmpty,
-                    );
+                        || parse_value_path(log_schema().host_key()).ok(),
+                        |k| k.path.clone(),
+                    )
+                    .map(LegacyKey::InsertIfEmpty);
```
Comment (lukesteensen) [inline src/sources/socket/mod.rs:239]: nitpick: Would this be less awkward if `log_schema().host_key()` returned a parsed path value? Maybe that's something we're already moving towards, not sure.

File: src/sources/socket/tcp.rs [hunk 3]
```diff
@@ -43,14 +46,14 @@ pub struct TcpConfig {
     /// By default, the [global `log_schema.host_key` option][global_host_key] is used.
     ///
     /// [global_host_key]: https://vector.dev/docs/reference/configuration/global-options/#log_schema.host_key
-    host_key: Option<String>,
+    host_key: Option<OptionalValuePath>,
```
Comment (lukesteensen) [inline src/sources/socket/tcp.rs:49]: `Option<Optional...>` is a little weird, but I don't have much context on this type.

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**